File: 2022/day-03/solution.js
// https://adventofcode.com/2022/day/3


const fs = require('fs');

function parseData() {
    try {
        return fs.readFileSync('./day-03/input.txt', 'utf-8').split('\n');
    } catch (err) {
        console.error(err);
    }
}


function problemOne(rucksacks) {
    let total = 0;

    for (let rucksack of rucksacks) {
        let mid = Math.floor(rucksack.length / 2);
        let left = [...rucksack.substring(0, mid)];
        let right = [...rucksack.substring(mid)];
        
        let similarItem = left.filter(value => right.includes(value))[0];
        let priority = similarItem.charCodeAt(0);

        if (priority > 96) {
            total += priority - 96;
        } else {
            total += priority - 38;
        }
    }

    return total;
}


function problemTwo(rucksacks) {
    const N = rucksacks.length / 3;
    let total = 0;

    for (let i = 0; i < rucksacks.length; i += 3) {
        let sack1 = [...rucksacks[i]]
        let sack2 = [...rucksacks[i + 1]]
        let sack3 = [...rucksacks[i + 2]]
        
        let similarItem = sack1.filter(value => {
            if (sack2.includes(value) && sack3.includes(value)) {
                return true;
            }
            return false;
        });

        let priority = similarItem[0].charCodeAt(0);

        if (priority > 96) {
            total += priority - 96;
        } else {
            total += priority - 38;
        }
    }

    return total;
}


const rucksacks = parseData();
console.log(`Problem One: ${problemOne(rucksacks)}`)
console.log(`Problem Two: ${problemTwo(rucksacks)}`)

File: 2022/day-03/solution.py
rucksacks = []

def main():
    read_file()
    print(problem_one())
    print(problem_two())


def read_file():
    with open("./day-03/input.txt") as f:
        for row in f:
            rucksacks.append(row)


def problem_one() -> int:
    sum = 0

    for rucksack in rucksacks:
        # Split each rucksack into two equal compartments
        n = int(len(rucksack) / 2)
        first_compartment = rucksack[0:n]
        second_compartment = rucksack[n:-1]

        for letter in first_compartment:
            if letter in second_compartment:
                # Do a bit of maths with the ascii values to get the priority
                if letter.islower():
                    sum += ord(letter) - 96
                    break
                else:
                    sum += ord(letter) - 38
                    break

    return sum


def problem_two() -> int:
    # Divide the rucksacks into three groups
    n = int(len(rucksacks) / 3)
    sum = 0
    index = 0

    for _ in range(n):
        for letter in rucksacks[index]:
            if letter in rucksacks[index + 1] and letter in rucksacks[index + 2]:
                if letter.islower():
                    sum += ord(letter) - 96
                    break
                else:
                    sum += ord(letter) - 38
                    break
        
        index += 3

    return sum
        

if __name__ == "__main__":
    main()
